libs/net/nodenetworkconfigurationpolicy.py

```diff
@@ -0,0 +1,250 @@
+                    yield nnce
+
+    def _process_failed_status(self, failed_condition_reason: str) -> None:
+        last_err_msg = None
+        for failed_nnce in self._get_failed_nnce():
+            nnce_name = failed_nnce.instance.metadata.name
+            nnce_dict = failed_nnce.instance.to_dict()
+            for cond in nnce_dict["status"]["conditions"]:
+                err_msg = self._get_nnce_error_msg(nnce_name=nnce_name, nnce_condition=cond)
+                if err_msg:
+                    last_err_msg = err_msg
+
+        raise NNCPConfigurationFailed(f"Reason: {failed_condition_reason}\n{last_err_msg}")
+
+    def wait_for_status_success(self) -> list[dict]:
+        failed_condition_reason = self.Conditions.Reason.FAILED_TO_CONFIGURE
+        no_match_node_condition_reason = self.Conditions.Reason.NO_MATCHING_NODE
+
+        samples = TimeoutSampler(
+            wait_timeout=TIMEOUT_90SEC,
+            sleep=TIMEOUT_5SEC,
+            func=lambda: self.instance.status.conditions,
+        )
+        for sample in samples:
+            if sample:
+                for nncp_condition in sample:
+                    if (
+                        nncp_condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED
+                        and nncp_condition["status"] == self.Condition.Status.TRUE
+                    ):
+                        self.logger.info(f"NNCP {self.name} configured Successfully")
+                        return sample
+
+                    elif nncp_condition["reason"] == no_match_node_condition_reason:
+                        raise NNCPConfigurationFailed(f"{self.name}. Reason: {no_match_node_condition_reason}")
+
+                    elif nncp_condition["reason"] == failed_condition_reason:
+                        self._process_failed_status(failed_condition_reason=failed_condition_reason)
+
+        raise NNCPConfigurationFailed(f"{self.name}. None of the conditions were met.")
+
+    def _absent_interface(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
+            for _iface in self.res["spec"]["desiredState"]["interfaces"]:
+                _iface["state"] = Resource.Interface.State.ABSENT
+
+            desired_state["interfaces"] = self.res["spec"]["desiredState"]["interfaces"]
+
+        ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
+
+    def clean_up(self, wait: bool = True, timeout: int | None = None) -> bool:
+        try:
+            self._absent_interface()
+            self.wait_for_status_success()
+        except Exception as exp:
+            self.logger.error(exp)
+
+        return self.delete()
+
+    def interfaces(self) -> list[dict]:
+        return self.res["spec"]["desiredState"]["interfaces"]
+
+
+def replace_underscore_with_hyphen_and_filter_none_values(data: list[tuple[str, Any]]) -> dict[str, Any]:
```

Comment:
2 things:
1. The name of this function is confusing. It implies that it only performs some cosmetic actions, but it actually produces a dict out of a list object.
2. The part of replacing underscores seems like the function is covering-up for the callers misbehavior. The caller should create a proper NNCP object, which also includes creating dict keys according to the kubernetes rules.

Reply:
1. The actual conversion is known from the caller (in to_dict), but the function doesn't care who calls it and should declare itself correctly. Ack, I'll rename.
2. Not exactly, the caller uses dataclass, which behave in a certain non-kuberneted way -> the field names (keys) can only include underscores and not hyphens. Later we need to alter it according to our kubernetes needs.

Reply:
2. Ack, thank you for explaining.